cpp/src/arrow/record_batch.cc

```diff
@@ -248,49 +248,100 @@ Result<std::shared_ptr<StructArray>> RecordBatch::ToStructArray() const {
-    out_values += batch.num_rows();
-  }  // End loop through columns
+    // If the column is of the same type than resulting data type
+    if (TypeTraits<DataType>::type_singleton() == batch.column(i)->type()) {
+      const auto* in_values = batch.column(i)->data()->GetValues<CType>(1);
+
+      memcpy(out_values, in_values, sizeof(CType) * length);
+      out_values += length;
+    } else {  // If the column is different type than resulting data type
+      switch (batch.column(i)->type_id()) {
+        case Type::HALF_FLOAT: {
+          auto* in_values = batch.column(i)->data()->GetValues<uint16_t>(1);
+          for (int64_t i = 0; i < length; ++i) {
+            *out_values++ = static_cast<CType>(*in_values++);
+          }
+          break;
+        }
+          TYPE_CASE(UINT8)
+          TYPE_CASE(UINT16)
+          TYPE_CASE(UINT32)
+          TYPE_CASE(UINT64)
+          TYPE_CASE(INT8)
+          TYPE_CASE(INT16)
+          TYPE_CASE(INT32)
+          TYPE_CASE(INT64)
+          TYPE_CASE(FLOAT)
+          TYPE_CASE(DOUBLE)
+        default:
+          break;
+      }
+    }
+  }
 }
 
+#undef TYPE_CASE
+
 Result<std::shared_ptr<Tensor>> RecordBatch::ToTensor(MemoryPool* pool) const {
   if (num_columns() == 0) {
     return Status::TypeError(
         "Conversion to Tensor for RecordBatches without columns/schema is not "
         "supported.");
   }
-  const auto& type = column(0)->type();
-  // Check for supported data types
-  if (!is_integer(type->id()) && !is_floating(type->id())) {
-    return Status::TypeError("DataType is not supported: ", type->ToString());
-  }
-  // Check for uniform data type
   // Check for no validity bitmap of each field
   for (int i = 0; i < num_columns(); ++i) {
     if (column(i)->null_count() > 0) {
       return Status::TypeError("Can only convert a RecordBatch with no nulls.");
     }
-    if (column(i)->type() != type) {
-      return Status::TypeError("Can only convert a RecordBatch with uniform data type.");
+  }
+
+  // Check for supported data types and merge fields
+  // to get the resulting uniform data type
+  if (!is_integer(column(0)->type()->id()) && !is_floating(column(0)->type()->id())) {
+    return Status::TypeError("DataType is not supported: ",
+                             column(0)->type()->ToString());
+  }
+  std::shared_ptr<DataType> result_type = column(0)->type();
+
+  if (num_columns() > 1) {
+    Field::MergeOptions options;
+    options.promote_integer_to_float = true;
+    options.promote_integer_sign = true;
+    options.promote_numeric_width = true;
+
+    for (int i = 1; i < num_columns(); ++i) {
+      if (!is_integer(column(i)->type()->id()) && !is_floating(column(i)->type()->id())) {
+        return Status::TypeError("DataType is not supported: ",
+                                 column(i)->type()->ToString());
+      }
+      result_type = *MergeTypes(result_type, column(i)->type(), options);
```

Comment:
Are we sure MergeTypes always succeeds? (probably yes because you already checked we only have integer or floating point data?) But might want to use a `ARROW_ASSIGN_OR_RAISE`